File: src-tauri/src/evaluation/line_evaluators.rs
use serde_json::{json, Value};
use std::collections::HashMap;

use super::errors::geometry_error;
use super::math::{
    angle_from_to, arc_tangent_angles, circle_through_three_points, normalize_degrees,
};
use super::numeric_expression::evaluate_numeric_or_push;
use super::point_anchor::{anchor_reference_element_id, computed_point, point_anchor_or_error};
use super::types::{element_id, element_name, insert_geometry, EvaluationState, Point};

pub(crate) fn evaluate_line(
    element: &Value,
    local_variables: &(HashMap<String, f64>, HashMap<String, String>),
    state: &mut EvaluationState,
) {
    let Some(start_anchor) = element.get("startPoint") else {
        return;
    };
    let Some(end_anchor) = element.get("endPoint") else {
        return;
    };
    let Some(start) = point_anchor_or_error(
        element,
        start_anchor,
        "start",
        state,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(end) = point_anchor_or_error(
        element,
        end_anchor,
        "end",
        state,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let dx = end.x - start.x;
    let dy = end.y - start.y;
    let length = dx.hypot(dy);
    let start_angle = angle_from_to(&start, &end);
    let end_angle = angle_from_to(&end, &start);
    let id = element_id(element).unwrap_or_default();
    insert_geometry(
        state,
        id.clone(),
        json!({
            "kind": "line",
            "elementId": id,
            "name": element_name(element),
            "startPointId": anchor_reference_element_id(start_anchor),
            "endPointId": anchor_reference_element_id(end_anchor),
            "start": computed_point(start.element_id, start.name, start.x, start.y),
            "end": computed_point(end.element_id, end.name, end.x, end.y),
            "length": length,
            "startAngleDeg": start_angle,
            "endAngleDeg": end_angle,
            "startTangentAngleDeg": start_angle,
            "endTangentAngleDeg": end_angle
        }),
    );
}

pub(crate) fn evaluate_arc_line(
    element: &Value,
    local_variables: &(HashMap<String, f64>, HashMap<String, String>),
    state: &mut EvaluationState,
) {
    let Some(center_anchor) = element.get("centerPoint") else {
        return;
    };
    let Some(center) = point_anchor_or_error(
        element,
        center_anchor,
        "center",
        state,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(radius) = evaluate_numeric_or_push(
        element.get("radius").unwrap_or(&Value::Null),
        state,
        element,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(start_angle_deg) = evaluate_numeric_or_push(
        element.get("startAngleDeg").unwrap_or(&Value::Null),
        state,
        element,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(end_angle_deg) = evaluate_numeric_or_push(
        element.get("endAngleDeg").unwrap_or(&Value::Null),
        state,
        element,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let safe_radius = if radius > 0.0 { radius } else { 0.0 };
    let sweep_angle_deg = normalize_degrees(end_angle_deg - start_angle_deg);
    let (start_tangent_angle_deg, end_tangent_angle_deg) =
        arc_tangent_angles(start_angle_deg, end_angle_deg, sweep_angle_deg);
    let id = element_id(element).unwrap_or_default();
    insert_arc_line_geometry(
        state,
        ArcGeometry {
            id,
            name: element_name(element),
            center_point_id: anchor_reference_element_id(center_anchor),
            center,
            radius,
            safe_radius,
            start_angle_deg,
            end_angle_deg,
            start_tangent_angle_deg,
            end_tangent_angle_deg,
            sweep_angle_deg,
        },
    );
}

pub(crate) fn evaluate_three_point_arc_line(
    element: &Value,
    local_variables: &(HashMap<String, f64>, HashMap<String, String>),
    state: &mut EvaluationState,
) {
    let Some(point1) = point_anchor_or_error(
        element,
        element.get("point1").unwrap_or(&Value::Null),
        "point1",
        state,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(point2) = point_anchor_or_error(
        element,
        element.get("point2").unwrap_or(&Value::Null),
        "point2",
        state,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(point3) = point_anchor_or_error(
        element,
        element.get("point3").unwrap_or(&Value::Null),
        "point3",
        state,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(start_angle_deg) = evaluate_numeric_or_push(
        element.get("startAngleDeg").unwrap_or(&Value::Null),
        state,
        element,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(end_angle_deg) = evaluate_numeric_or_push(
        element.get("endAngleDeg").unwrap_or(&Value::Null),
        state,
        element,
        &local_variables.0,
        &local_variables.1,
    ) else {
        return;
    };
    let Some(circle) = circle_through_three_points(&point1, &point2, &point3) else {
        state.errors.push(geometry_error(
            element,
            format!(
                "{} は点1・点2・点3から円を作れません。3点が重複しているか、一直線上にあります。別の3点を指定してください。",
                element_name(element)
            ),
        ));
        return;
    };

    let sweep_angle_deg = normalize_degrees(end_angle_deg - start_angle_deg);
    let (start_tangent_angle_deg, end_tangent_angle_deg) =
        arc_tangent_angles(start_angle_deg, end_angle_deg, sweep_angle_deg);
    let id = element_id(element).unwrap_or_default();
    insert_arc_line_geometry(
        state,
        ArcGeometry {
            id: id.clone(),
            name: element_name(element),
            center_point_id: None,
            center: Point {
                element_id: format!("{id}:center"),
                name: format!("{}.中心点", element_name(element)),
                x: circle.x,
                y: circle.y,
            },
            radius: circle.radius,
            safe_radius: circle.radius,
            start_angle_deg,
            end_angle_deg,
            start_tangent_angle_deg,
            end_tangent_angle_deg,
            sweep_angle_deg,
        },
    );
}

struct ArcGeometry {
    id: String,
    name: String,
    center_point_id: Option<String>,
    center: Point,
    radius: f64,
    safe_radius: f64,
    start_angle_deg: f64,
    end_angle_deg: f64,
    start_tangent_angle_deg: f64,
    end_tangent_angle_deg: f64,
    sweep_angle_deg: f64,
}

fn insert_arc_line_geometry(state: &mut EvaluationState, arc: ArcGeometry) {
    let start_angle_rad = arc.start_angle_deg.to_radians();
    let end_angle_rad = arc.end_angle_deg.to_radians();
    insert_geometry(
        state,
        arc.id.clone(),
        json!({
            "kind": "arcLine",
            "elementId": arc.id,
            "name": arc.name,
            "centerPointId": arc.center_point_id,
            "center": computed_point(arc.center.element_id, arc.center.name, arc.center.x, arc.center.y),
            "start": computed_point(format!("{}:start", arc.id), format!("{}.始点", arc.name), arc.center.x + start_angle_rad.cos() * arc.safe_radius, arc.center.y + start_angle_rad.sin() * arc.safe_radius),
            "end": computed_point(format!("{}:end", arc.id), format!("{}.終点", arc.name), arc.center.x + end_angle_rad.cos() * arc.safe_radius, arc.center.y + end_angle_rad.sin() * arc.safe_radius),
            "radius": arc.radius,
            "startAngleDeg": arc.start_angle_deg,
            "endAngleDeg": arc.end_angle_deg,
            "startTangentAngleDeg": arc.start_tangent_angle_deg,
            "endTangentAngleDeg": arc.end_tangent_angle_deg,
            "sweepAngleDeg": arc.sweep_angle_deg,
            "length": arc.safe_radius * arc.sweep_angle_deg.to_radians()
        }),
    );
}

File: src/geometry/lineEvaluators.ts
import type { CadElement, ComputedPoint } from "../types/geometry";
import { anchorReferenceElementId } from "../model/pointAnchors";
import {
  approximateBezierSegmentLength,
  CIRCLE_EPSILON,
  handlePoint,
  normalizeDegrees
} from "./evaluateGeometryPrimitives";
import { dependencyError, geometryError, getPointAnchorOrError, numericError } from "./evaluationContext";
import type { ElementEvaluationContext } from "./elementEvaluatorTypes";
import { lineTangentAngles } from "./lineMeasurements";
import { arcGeometryKernel, commonTangentGeometryKernel, polarLineGeometryKernel, segmentGeometryKernel, throughArcGeometryKernel } from "./geometryValueKernels";
import { resolveLineGeometryInput } from "./lineGeometryInput";

export const evaluateLineElement = (element: CadElement, context: ElementEvaluationContext) => {
  const {
    computedGeometry,
    elementsById,
    errors,
    disabledByGroupId,
    computedGeometryValues,
    localVariables: { localVariableValues, localVariableNames }
  } = context;

  switch (element.type) {
      case "line": {
        const start = getPointAnchorOrError(
          element,
          element.startPoint,
          "start",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames,
          disabledByGroupId,
          undefined,
          computedGeometryValues
        );
        const end = getPointAnchorOrError(
          element,
          element.endPoint,
          "end",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames,
          disabledByGroupId,
          undefined,
          computedGeometryValues
        );
        if (!start || !end) {
          break;
        }

        const structural = segmentGeometryKernel(start, end);
        computedGeometry.set(element.id, {
          ...structural,
          elementId: element.id,
          name: element.name,
          startPointId: anchorReferenceElementId(element.startPoint),
          endPointId: anchorReferenceElementId(element.endPoint),
          start: { kind: "point", elementId: start.elementId, name: start.name, x: start.x, y: start.y },
          end: { kind: "point", elementId: end.elementId, name: end.name, x: end.x, y: end.y }
        });
        break;
      }
      case "polyline": {
        const minimumPointCount = element.closed ? 3 : 2;
        if (element.points.length < minimumPointCount) {
          errors.push(geometryError(
            element,
            `${element.name} の点数が不足しています。${element.closed ? "閉じた折れ線には3点以上" : "開いた折れ線には2点以上"}の点を指定してください。`
          ));
          break;
        }

        const points = element.points.map((point, index) => getPointAnchorOrError(
          element,
          point,
          `point${index + 1}`,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames,
          disabledByGroupId,
          undefined,
          computedGeometryValues
        ));
        if (points.some((point) => !point)) break;
        const resolvedPoints = points as ComputedPoint[];
        const segments = resolvedPoints.slice(0, -1).map((start, index) => {
          const end = resolvedPoints[index + 1]!;
          return {
            kind: "line" as const,
            start,
            end,
            length: Math.hypot(end.x - start.x, end.y - start.y)
          };
        });
        const first = resolvedPoints[0]!;
        const last = resolvedPoints.at(-1)!;
        if (element.closed && Math.hypot(last.x - first.x, last.y - first.y) > CIRCLE_EPSILON) {
          segments.push({
            kind: "line",
            start: last,
            end: first,
            length: Math.hypot(first.x - last.x, first.y - last.y)
          });
        }
        const nonZero = segments.filter((segment) => segment.length > CIRCLE_EPSILON);
        const startTangentAngleDeg = nonZero[0]
          ? lineTangentAngles(nonZero[0].start, nonZero[0].end).startTangentAngleDeg
          : null;
        const lastNonZero = nonZero.at(-1);
        const endTangentAngleDeg = lastNonZero
          ? lineTangentAngles(lastNonZero.start, lastNonZero.end).endTangentAngleDeg
          : null;
        computedGeometry.set(element.id, {
          kind: "polyline",
          elementId: element.id,
          name: element.name,
          segments,
          closed: element.closed,
          start: first,
          end: element.closed ? first : last,
          length: segments.reduce((sum, segment) => sum + segment.length, 0),
          startTangentAngleDeg,
          endTangentAngleDeg
        });
        break;
      }
      case "angleLengthLine": {
        const start = getPointAnchorOrError(
          element,
          element.startPoint,
          "start",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames,
          disabledByGroupId,
          undefined,
          computedGeometryValues
        );
        if (!start) {
          break;
        }

        const angleDeg = numericError(
          element,
          element.angleDeg,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        const length = numericError(
          element,
          element.length,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        if (angleDeg === undefined || length === undefined) {
          break;
        }

        const structural = polarLineGeometryKernel(start, angleDeg, length);
        const end: ComputedPoint = {
          kind: "point",
          elementId: `${element.id}:end`,
          name: `${element.name}.終点`,
          ...structural.end
        };
        computedGeometry.set(element.id, {
          kind: "line",
          elementId: element.id,
          name: element.name,
          startPointId: anchorReferenceElementId(element.startPoint),
          endPointId: null,
          start: {
            kind: "point",
            elementId: `${element.id}:start`,
            name: `${element.name}.始点`,
            x: start.x,
            y: start.y
          },
          end,
          length: structural.length,
          startAngleDeg: structural.startAngleDeg,
          endAngleDeg: structural.endAngleDeg,
          startTangentAngleDeg: structural.startTangentAngleDeg,
          endTangentAngleDeg: structural.endTangentAngleDeg
        });
        break;
      }
      case "commonTangentLine": {
        const firstGeometry = resolveLineGeometryInput(context, "firstLineId", element.firstLineId);
        const secondGeometry = resolveLineGeometryInput(context, "secondLineId", element.secondLineId);
        if (!firstGeometry) {
          errors.push(dependencyError(element, element.firstLineId, elementsById, disabledByGroupId, errors));
        } else if (firstGeometry.kind !== "arcLine") {
          errors.push(geometryError(element, "first に円弧が指定されていません。共通接線には円弧を指定してください。"));
        }
        if (!secondGeometry) {
          errors.push(dependencyError(element, element.secondLineId, elementsById, disabledByGroupId, errors));
        } else if (secondGeometry.kind !== "arcLine") {
          errors.push(geometryError(element, "second に円弧が指定されていません。共通接線には円弧を指定してください。"));
        }
        if (firstGeometry?.kind !== "arcLine" || secondGeometry?.kind !== "arcLine") break;

        const result = commonTangentGeometryKernel(firstGeometry, secondGeometry, element.kind, element.side);
        if ("errors" in result) {
          for (const message of result.errors) errors.push(geometryError(element, message));
          break;
        }
        const { line: structural } = result;
        const start: ComputedPoint = {
          kind: "point",
          elementId: `${element.id}:start`,
          name: `${element.name}.始点`,
          ...structural.start
        };
        const end: ComputedPoint = {
          kind: "point",
          elementId: `${element.id}:end`,
          name: `${element.name}.終点`,
          ...structural.end
        };
        computedGeometry.set(element.id, {
          kind: "line",
          elementId: element.id,
          name: element.name,
          startPointId: null,
          endPointId: null,
          start,
          end,
          length: structural.length,
          startAngleDeg: structural.startAngleDeg,
          endAngleDeg: structural.endAngleDeg,
          startTangentAngleDeg: structural.startTangentAngleDeg,
          endTangentAngleDeg: structural.endTangentAngleDeg
        });
        break;
      }
      case "arcLine": {
        const center = getPointAnchorOrError(
          element,
          element.centerPoint,
          "center",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames,
          disabledByGroupId,
          undefined,
          computedGeometryValues
        );
        if (!center) {
          break;
        }

        const radius = numericError(
          element,
          element.radius,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames,
          undefined,
          undefined,
          computedGeometryValues
        );
        const startAngleDeg = numericError(
          element,
          element.startAngleDeg,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames,
          undefined,
          undefined,
          computedGeometryValues
        );
        const endAngleDeg = numericError(
          element,
          element.endAngleDeg,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames,
          undefined,
          undefined,
          computedGeometryValues
        );
        if (radius === undefined || startAngleDeg === undefined || endAngleDeg === undefined) {
          break;
        }

        if (!(radius > 0)) {
          errors.push(geometryError(element, `${element.name} の半径は0より大きい値で指定してください。`));
          break;
        }

        const structural = arcGeometryKernel(
          { x: center.x, y: center.y },
          radius,
          startAngleDeg,
          endAngleDeg,
          element.direction ?? "counterclockwise"
        );
        computedGeometry.set(element.id, {
          ...structural,
          elementId: element.id,
          name: element.name,
          centerPointId: anchorReferenceElementId(element.centerPoint),
          center,
          start: {
            kind: "point",
            elementId: `${element.id}:start`,
            name: `${element.name}.始点`,
            ...structural.start
          },
          end: {
            kind: "point",
            elementId: `${element.id}:end`,
            name: `${element.name}.終点`,
            ...structural.end
          }
        });
        break;
      }
      case "threePointArcLine": {
        const point1 = getPointAnchorOrError(
          element,
          element.point1,
          "point1",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
          , undefined,
          undefined,
          computedGeometryValues
        );
        const point2 = getPointAnchorOrError(
          element,
          element.point2,
          "point2",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
          , undefined,
          undefined,
          computedGeometryValues
        );
        const point3 = getPointAnchorOrError(
          element,
          element.point3,
          "point3",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        if (!point1 || !point2 || !point3) {
          break;
        }

        const startAngleDeg = numericError(
          element,
          element.startAngleDeg,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        const endAngleDeg = numericError(
          element,
          element.endAngleDeg,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        if (startAngleDeg === undefined || endAngleDeg === undefined) {
          break;
        }

        const structural = throughArcGeometryKernel(
          { x: point1.x, y: point1.y },
          { x: point2.x, y: point2.y },
          { x: point3.x, y: point3.y },
          startAngleDeg,
          endAngleDeg
        );
        if (!structural) {
          errors.push(
            geometryError(
              element,
              `${element.name} は点1・点2・点3から円を作れません。3点が重複しているか、一直線上にあります。別の3点を指定してください。`
            )
          );
          break;
        }
        computedGeometry.set(element.id, {
          kind: "arcLine",
          elementId: element.id,
          name: element.name,
          centerPointId: null,
          center: {
            kind: "point",
            elementId: `${element.id}:center`,
            name: `${element.name}.中心点`,
            x: structural.center.x,
            y: structural.center.y
          },
          start: {
            kind: "point",
            elementId: `${element.id}:start`,
            name: `${element.name}.始点`,
            ...structural.start
          },
          end: {
            kind: "point",
            elementId: `${element.id}:end`,
            name: `${element.name}.終点`,
            ...structural.end
          },
          radius: structural.radius,
          startAngleDeg: structural.startAngleDeg,
          endAngleDeg: structural.endAngleDeg,
          startTangentAngleDeg: structural.startTangentAngleDeg,
          endTangentAngleDeg: structural.endTangentAngleDeg,
          sweepAngleDeg: structural.sweepAngleDeg,
          length: structural.length
        });
        break;
      }
      case "bezierCurve": {
        const start = getPointAnchorOrError(
          element,
          element.startPoint,
          "start",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        const end = getPointAnchorOrError(
          element,
          element.endPoint,
          "end",
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        const intermediatePoints = element.intermediatePoints.map((intermediate) =>
          getPointAnchorOrError(
            element,
            intermediate.point,
            `intermediate:${intermediate.id}`,
            computedGeometry,
            elementsById,
            errors,
            localVariableValues,
            localVariableNames,
            disabledByGroupId,
            undefined,
            computedGeometryValues
          )
        );
        if (!start || !end || intermediatePoints.some((point) => !point)) {
          break;
        }

        const startHandleAngleDeg = numericError(
          element,
          element.startHandleAngleDeg,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        const startHandleLength = numericError(
          element,
          element.startHandleLength,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        const endHandleAngleDeg = numericError(
          element,
          element.endHandleAngleDeg,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        const endHandleLength = numericError(
          element,
          element.endHandleLength,
          computedGeometry,
          elementsById,
          errors,
          localVariableValues,
          localVariableNames
        );
        const intermediateHandles = element.intermediatePoints.map((intermediate) => ({
          angleDeg: numericError(
            element,
            intermediate.handleAngleDeg,
            computedGeometry,
            elementsById,
            errors,
            localVariableValues,
            localVariableNames
          ),
          incomingLength: numericError(
            element,
            intermediate.incomingHandleLength,
            computedGeometry,
            elementsById,
            errors,
            localVariableValues,
            localVariableNames
          ),
          outgoingLength: numericError(
            element,
            intermediate.outgoingHandleLength,
            computedGeometry,
            elementsById,
            errors,
            localVariableValues,
            localVariableNames
          )
        }));
        if (
          startHandleAngleDeg === undefined ||
          startHandleLength === undefined ||
          endHandleAngleDeg === undefined ||
          endHandleLength === undefined ||
          intermediateHandles.some(
            (handle) =>
              handle.angleDeg === undefined ||
              handle.incomingLength === undefined ||
              handle.outgoingLength === undefined
          )
        ) {
          break;
        }

        const anchors = [start, ...(intermediatePoints as ComputedPoint[]), end];
        const outgoingHandles = [
          handlePoint(start, startHandleAngleDeg, startHandleLength),
          ...intermediateHandles.map((handle, index) =>
            handlePoint(anchors[index + 1], handle.angleDeg!, handle.outgoingLength!)
          )
        ];
        const incomingHandles = [
          ...intermediateHandles.map((handle, index) =>
            handlePoint(anchors[index + 1], handle.angleDeg! + 180, handle.incomingLength!)
          ),
          handlePoint(end, endHandleAngleDeg + 180, endHandleLength)
        ];
        const segments = anchors.slice(0, -1).map((anchor, index) => ({
          startPointId: anchor.elementId,
          endPointId: anchors[index + 1].elementId,
          start: anchor,
          control1: outgoingHandles[index],
          control2: incomingHandles[index],
          end: anchors[index + 1]
        }));

        computedGeometry.set(element.id, {
          kind: "bezierCurve",
          elementId: element.id,
          name: element.name,
          startPointId: anchorReferenceElementId(element.startPoint),
          endPointId: anchorReferenceElementId(element.endPoint),
          intermediatePointIds: element.intermediatePoints.flatMap((point) =>
            anchorReferenceElementId(point.point) ? [anchorReferenceElementId(point.point)!] : []
          ),
          intermediateSlotIds: element.intermediatePoints.map((point) => point.id),
          segments,
          length: segments.reduce(
            (sum, segment) => sum + approximateBezierSegmentLength(segment),
            0
          ),
          startTangentAngleDeg: normalizeDegrees(startHandleAngleDeg),
          endTangentAngleDeg: normalizeDegrees(endHandleAngleDeg + 180),
          startHandleAngleDeg,
          startHandleLength,
          endHandleAngleDeg,
          endHandleLength
        });
        break;
      }

    default:
      return false;
  }
  return true;
};
